accepted/00097-recurring-actions.md

```diff
@@ -0,0 +1,193 @@
+
+Since it not currently possible and the number of recurring action schedules is expected to grow significantly we also want to support being able to filter the list as well as implement server side pagination.
+
+On the system level the `Recurring States` tab is currently displayed as a sub to of the `States` tab. It will be moved to the top level tabs and renamed to `Recurring Actions` since it's no longer going to be state specific.
+
+For action type specific frontend implementations see below.
+
+## Specific implementations
+
+### Custom States (Config Channels)
+
+The idea here is to schedule a `Apply states[custom]` action on a recurring basis. The states executed depend on what is configured in `States` -> `Configuration Channels` on System / Group / Org level.
+
+Since which states to run is preconfigured for each entity type we don't actually need a new parameter in `suseRecurringActionType` table.
+
+The backend part can be implemented simply following the workflow described above.
+
+For the frontend I'd suggest to add a `Create Recurring` button on the `Configuration Channels` page on System / Group / Org level similar as it was done for `Highstate`.
+The frontend pages for the highstate scheduling can be reused as they are and just pass/receive a `customstate` action type to/from the Controller class.
+
+### Content Lifecycle Management
+
+Recurring CLM management will work a bit different compared to other recurring action types since we don't have an entity here for which an action will be scheduled.
+
+To make it work I propose to create a new entity that could be called `NoEntityRecurringAction` it would set a new `target_type` `no-entity` in the `suseRecurringAction` table and still use the `org_id` as identifier since CLM project are bound to organizations.
+
+In contrast to the `OrgRecurringAction` this type would not calculate the orgs minions though since it's not required here.
+
+Two new action types `clm-build` and `clm-promote` would be supported (including `RecurringContentBuildType` and `RecurringContentPromoteType` entities).
+
+The `suseRecurringActionType` table would store a reference to the project (project_id) for the `clm-build` type and in addition a reference to the environment (env_id) to promote for the `clm-promote` type.
+
+The `RecurringActionApplyJob` `Taskomatic` job would then call the functions to build/promote a CLM project on recurrent basis.
+
+On the frontend side I'd suggest to add a new main menu item called e.g. `Recurring Lifecycle Management` in `Content Lifecycle` (below `Filters`).
+
+Here a list with existing recurring CLM build/promote schedules would be displayed as well as a button to prompt to create new schedules.
+
+The create page would add new input fields to select the project, the type (build/promote) and the environment to promote (if chosen).
+
+### Action Chains
```

Comment:
I'm not sure about this one, and it would need further investigation.
If we allow the user to define a list of stalt states to run, and also their order, that can be seen more or less as an action chain.

Reply:
As MC explained it to me the justification for action chains is that salt states don't work with a reboot in between. Also I guess there are customers that want to strictly stick to the web UI for their workflows. For them it would be nice if they'd be able to 'click' their own reusable/automated workflow together on the UI.

But I agree this needs further investigation

Reply:
Right. I think Action Chains should be the "last" to implemented and it require a re-design of Action Chains.
I would propose to only think about 5.0 or later where "traditional management" is removed.
Then we can think of removing the UI parts of action chains from the single pages like package install/update/delete, etc.
We need to generate states from UI and we would need new UIs for it.
This is a new and independent RFC.

Reply:
I agree with that. But in that case, we need to be careful with what we implement now in recurrent actions.
If we allow users to define a list of states to be applied and in which order, that will be similar to action chains, with the exception of the restart. This means we should not allow the restart to be added or it can fail to execute.
This lead to other another question: should we allow a list of states (and its order) to be defined or just one single state?

If we allow the list of states, that will be overlapping with action chains and make the implementation hard for this first stage.

We could say that in the phase we only support one single state (can be a configuration channel, some of the internal suma states, or a free text) and then when we have recurrent action channel, the user could define the list of states/actions to be applied.